osu.Game.Rulesets.Catch/CatchInputManager.cs

```diff
@@ -13,6 +18,26 @@ public CatchInputManager(RulesetInfo ruleset)
             : base(ruleset, 0, SimultaneousBindingMode.Unique)
         {
         }
+
+        public override void HandleInputStateChange(InputStateChangeEvent inputStateChange)
+        {
+            if (inputStateChange is ReplayStateChangeEvent<CatchAction> replayStateChange)
+            {
+                var replayState = (RulesetInputManagerInputState<CatchAction>)replayStateChange.State;
+                var lastState = (CatchReplayState)replayState.LastReplayState;
+
+                if (lastState.CatcherX != null)
+                {
+                    foreach (Drawable drawable in NonPositionalInputQueue)
+                    {
+                        if (drawable is Catcher catcher)
+                            catcher.UpdatePosition(lastState.CatcherX.Value);
+                    }
+                }
+            }
```

Comment:
Yeah I don't know about this. This is all very ad-hoc. I can't see osu! (the ruleset) doing anything similar anywhere, and this looks pretty bad all things considered.

I feel this should be solved in a more reusable manner for the sake of the overall ruleset API instead of scraping all over the current structure. This'll probably break immediately if we have to change anything.

I'd be in favour of more abstraction here, definitely. Have you tried anything in that direction? If so, do you have any branches to show?

Reply:
That I said standard ruleset doing a similar thing was my confusion involved, but the example is `DrawableOsuHitObject` uses `IsHovered` (set by `InputManager` after mouse move input) (it is similar in the sense that it is a data flow from `InputManager` to descendants other than via `KeyBindingContainer`).

As for abstraction, I don't have an idea of any good abstraction currently. I also find the whole `RulesetInputManagerInputState`, `ReplayState`, etc. quite confusing, and probably reconsidering it is desired too.

I can start thinking about how to improve the API. However, **the issue solved here is quite a blatant correctness issue**, and I think solving it should be prioritized over such an API design thinking.

Reply:
I'm gonna push back on that last part a little - if this is a problem for catch, it may well be a problem for another custom ruleset, and as such I'm of the opinion that we should be looking to get a "nice" solution in as not to set a bad precedent.

Regardless I'll have a more detailed look at if this can be improved in any way. Not a fan of delaying code quality matters for "later", because "later" more often than not becomes "never".